### PR TITLE
binomial(n,k<0) is no longer always 0

### LINKED ISSUE #14577: Alternative definition of binomial coefficients for negative arguments and associated documentation
According to this paper (https://arxiv.org/pdf/1105.3689.pdf), if we define binomial coefficients `binomial(n,k)` in terms of gamma functions, then it follows formula (2.1) of the paper for negative integer n and integer k, that for example gives binomial(-1,-1) as 1 in contradiction with SymPy that gives 0. This is the definition taken by the Wolfram Language (http://mathworld.wolfram.com/BinomialCoefficient.html). Here is how the computation is done (https://www.johndcook.com/blog/computing_binomial_coefficients/). Should we follow this definition, since we are defining binomial with gamma functions?

SymPy currently defines for the sake of convenience for negative integer 'k' that this function will return zero no matter what valued is the other argument. Why? Can this be more precisely defined as the definition in http://functions.wolfram.com/GammaBetaErf/Binomial/02/ ?

### LINKED ISSUE #14612: Wrong combsimp simplifications
The current definition of binomials implemented in SymPy does not satisfy the symmetry identity `binomial(n, k) = binomial(n, n - k)` for any `n` and `k` (for example with `n = k = -1` is not true), so the following result of combsimp is wrong:
``` python
>>> n,k = symbols("n k")
>>> combsimp(binomial(n,k)-binomial(n,n-k))
0
```
This error is difficult to solve without implementing the definition of binomial proposed in issue #14577, since the error comes from the following rewrite that happens in combsimp:
``` python
>>> n,k = symbols("n k")
>>> (binomial(n,k)-binomial(n,n-k)).rewrite(gamma)
0
```
*Similar problem*
The current definition of binomials implemented in SymPy does not satisfy the following identity `rf(n,k) = gamma(k+1) * binomial(n+k-1,k)` for any `n` and `k` (for example with `n = 1, k = -1` is not true), so the following result of combsimp is wrong:
``` python
>>> n,k = symbols("n k")
>>> combsimp(rf(n,k) - gamma(k+1) * binomial(n+k-1,k))
0
```
Again, this error is difficult to solve without implementing the definition of binomial proposed in issue #14577, since the error comes from the following rewrite that happens in combsimp:
``` python
>>> n,k = symbols("n k")
>>> (rf(n,k) - gamma(k+1) * binomial(n+k-1,k)).rewrite(gamma)
0
```
This issue depends on the decision taken in issue #14577.

### LINKED ISSUE #19030: values passed to binomial().rewrite(gamma) || (factorial) don't agree with binomial()
values passed to binomial().rewrite(gamma) binomial().rewrite(factorial) don't agree with binomial() when substituted with values after rewriting
```python3
>>> n, k = symbols('n k', integer=True)
>>> for f in (gamma, factorial):
...        for i in range(-3,4):
...             for j in range(-3,4):
...                     if not Eq(binomial(n,k), binomial(n,k).rewrite(f)).subs({n:i,k:j}):
...                             f, i , j , binomial(n,k).subs({n:i,k:j}), binomial(n,k).rewrite(f).subs({n:i,k:j})
... 
(gamma, -3, 1, -3, nan)
(gamma, -3, 2, 6, nan)
(gamma, -3, 3, -10, nan)
(gamma, -2, 1, -2, nan)
(gamma, -2, 2, 3, nan)
(gamma, -2, 3, -4, nan)
(gamma, -1, 1, -1, nan)
(gamma, -1, 2, 1, nan)
(gamma, -1, 3, -1, nan)
(factorial, -3, 1, -3, nan)
(factorial, -3, 2, 6, nan)
(factorial, -3, 3, -10, nan)
(factorial, -2, 1, -2, nan)
(factorial, -2, 2, 3, nan)
(factorial, -2, 3, -4, nan)
(factorial, -1, 1, -1, nan)
(factorial, -1, 2, 1, nan)
(factorial, -1, 3, -1, nan)
```
that is
```python3
>>> binomial(n,k).subs({n:-1,k:3})
-1
>>> binomial(n,k).rewrite(gamma).subs({n:-1,k:3})
nan
```
is this is intended output or a side effect ?
I don't think this issue has been completely solved [#13890](https://github.com/sympy/sympy/issues/13980)

### LINKED ISSUE #19082: binomial._eval_Mod problems for composite modulus
```python
>>> binomial(5, 3)
10
>>> [binomial(5,3)%i for i in range(9,18)]
[1, 0, 10, 10, 10, 10, 10, 10, 10]
>>> [binomial(5, 3, evaluate=False) % i for i in range(9,18)]
[3, 0, 10, 6, 10, 2, 0, 14, 10]
>>> list(range(9,18))
[9, 10, 11, 12, 13, 14, 15, 16, 17]
```
It appears to only be correct when the binomial value equals the modulus or when the modulus is prime.

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

Fixes #19082
Fixes #19030
Fixes #14577
Fixes #14612
work of #18960 continues from this by defining `multinomial` for arbitrary input.

#### Brief description of what is fixed or changed

A new branch of behavior has been opened up for binomial(n,k).

The former behavior was to return 0 whenever k was negative. Now

`(n, k) == (n, n - k)` for all `n` and `k` but the identity `(n, k) = (n-1, k-1) + (n-1, k)` no longer holds for (0,0).
```python
>>> binomial(0,0)
1
>>> binomial(-1,-1),binomial(-1,0)
(1, 1)
```
In addition, all rewriting now has to check that the sign of
the arguments will produce a rewrite value that is true for
all values matching the assumptions of the arguments. As a case
in point, consider the Piecewise form for binomial(x,y)
```python
    def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
        from sympy import gamma, Abs, Lt
        isint = lambda x: Eq(x, floor(x))
        nneg = lambda x: Or(Eq(x, 0), Eq(x/Abs(x), 1))
        for i in igen():
            if not n.has(i) and not k.has(i):
                break
        return Piecewise(
            (factorial(n)/factorial(k)/factorial(n - k),
                And(isint(n), isint(k),
                nneg(n), nneg(k), nneg(n - k))),
            (ff(n, k)/factorial(k),
                And(isint(k), nneg(k))),
            (ff(n, n - k)/factorial(n - k),
                And(isint(n - k), nneg(n - k))),
            (0, And(*[i for x in (n, k, n - k)
                for i in [isint(x), Lt(x, 0)]])),
            (gamma(n + 1)/gamma(k + 1)/gamma(n - k + 1),
                True))
```
The checks for non-negative values are in place to guarantee that
the given expression will be true for arguments with matching
assumptions. The 4th condition (x, y and x - y all negative integers)
is there to guard from returning gamma for that case which would give
nan when the actual answer is 0.

#### Release Notes

* functions
  * binomial(n,k) no longer gives 0 for negative values of k
  * binomial can now be rewritten as Piecewise and all other rewrites respect assumptions on args parsing

### CODE REVIEW

**Comment by asmeurer** [comment]
My main concern with changing the definition of binomial is how it affects simplification. Certain identities are not true for certain definitions, which means we shouldn't apply them in simplification unless the assumptions are met. Which identities are the most important in simplification, and for which definition(s) do they hold?

**Comment by iammosespaulr** [comment]
> Which identities are the most important in simplification, and for which definition(s) do they hold?

@asmeurer if you take a look below, you'll see that the first 5 identities hold without any assumptions (except `finite=True`)
https://github.com/sympy/sympy/blob/457fd213b4d17ce90c7db664e7948a4d472400ba/sympy/functions/combinatorial/tests/test_comb_factorials.py#L423-L437

---

w.r.t simplification, it tends to call `rewrite_as_gamma` and we've got that covered. It'll work given the correct assumptions.

File: sympy/series/tests/test_formal.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 def test_rational_algorithm():
     f = 1 / ((x - 1)**2 * (x - 2))
     assert rational_algorithm(f, x, k) == \
-        (-2**(-k - 1) + 1 - (factorial(k + 1) / factorial(k)), 0, 0)
+        (-2**(-k - 1) - k, 0, 0)
```
Comment (iammosespaulr) [inline sympy/series/tests/test_formal.py:19]: It actually simplified this part even further.

Comment (iammosespaulr) [reply sympy/series/tests/test_formal.py:19]: They're equivalent btw
```python3
>>> (-2**(-k - 1) + 1 - (factorial(k + 1) / factorial(k))).equals(-2**(-k - 1) - k)
True
```

**Comment by asmeurer** [comment]
Another thing that I'd like to see is documentation that discusses the definition, what the advantages are, alternate definitions (it should at least be clear that there are alternate definitions), and so on. It can go in the docstring, unless it ends up being really long, in which case it may be better as a separate document in the Sphinx docs.

**Comment by asmeurer** [comment]
And we should make it clear in the release notes that this is a breaking change.

File: sympy/functions/combinatorial/factorials.py [hunk 2]
```diff
@@ -868,192 +922,304 @@ def fdiff(self, argindex=1):
...
+                from sympy.core.compatibility import gmpy
+                return Ntype(gmpy.bincoef(n, k))
+
+            d, result = n - k, 1
+            for i in range(1, k + 1):
+                d += 1
+                result = result * d // i
+            return Ntype(result)
+        else:
+            return Mul(*[n + i for i in range(1 - k, 1)])/factorial(k)
```
Comment (iammosespaulr) [inline sympy/functions/combinatorial/factorials.py:948]: Screen Shot 2020-04-22 at 1 56 45 AM
[link here](https://en.wikipedia.org/wiki/Binomial_coefficient#Binomial_coefficients_as_polynomials)
This is what our implementation does.
```python
>>> n = -3; k = -5
>>> ff(n,k)/factorial(k)
nan
>>> k = n - k
>>> ff(n,k)/factorial(k) <- we choose the second option
6
```
and yes [`binomial(-3, -5) == 6`](http://www.wolframalpha.com/input/?i=Binomial%5B-3%2C%20-5%5D)

**Comment by iammosespaulr** [comment]
@smichr the sympy-bot only has one author in it's release notes, even though the commits have 2 authors. Bug ?

**Comment by asmeurer** [comment]
Yeah, the SymPy Bot doesn't know about the co-authored-by flag. I opened an issue here https://github.com/sympy/sympy-bot/issues/85. If it isn't fixed before this is merged you can just update the entry in the wiki manually.

**Comment by smichr** [comment]
> this is a breaking change.

@iammosespaulr , correct me if I am wrong, but the only thing that will break is the unlimited rewriting of binomial when assumptions do not allow. So what this means is that one must now declare their `k` values as positive to get the previous rewriting behavior. (I say positive because some rewriting won't happen if both `n` and `k` can be zero.)

**Comment by asmeurer** [comment]
> If it isn't fixed before this is merged you can just update the entry in the wiki manually.

@smichr you could also fix this by making the author metadata in the first commit point to @iammosespaulr. Git's own metadata can only have one author, which is why the co-authored-by thing in the commit message exists.
 
> @iammosespaulr , correct me if I am wrong, but the only thing that will break is the unlimited rewriting of binomial when assumptions do not allow. So what this means is that one must now declare their k values as positive to get the previous rewriting behavior. (I say positive because some rewriting won't happen if both n and k can be zero.)

I mean breaking in the sense that the definition of the function is different. So if someone relied on the old definition somehow, then this will break things for them. It might be worth in the documentation, in the part that discusses alternate definitions, to give Piecewise expressions for them for anyone who prefers to work with them.

**Comment by sylee957** [comment]
Should this be considered as fixing the wrong result or making a backward incompatible change?
I think that this an ambiguous place between, but I'm afraid that we may need to provide both binomial definitions with a new function like `binomial2` if we take backward compatibility most seriously.

On the contrary, we should investigate if it is easy to build the definition of Feller/Comtet/Knuth on top of this. I suppose that it can be built with a simple piecewise wrapper, and this can be documented.

**Comment by iammosespaulr** [comment]
> > this is a breaking change.
> 
> @iammosespaulr , correct me if I am wrong, but the only thing that will break is the unlimited rewriting of binomial when assumptions do not allow.

Yes!
> So what this means is that one must now declare their `k` values as positive to get the previous rewriting behavior. (I say positive because some rewriting won't happen if both `n` and `k` can be zero.)

Yeah!, `nonnegative` will do, cause we have `gamma(n + 1)` and while n is `nonnegative` we're good to go 😄

**Comment by iammosespaulr** [comment]
> It might be worth in the documentation, in the part that discusses alternate definitions, to give Piecewise expressions for them for anyone who prefers to work with them.

Gotcha

**Comment by iammosespaulr** [comment]
> Should this be considered as fixing the wrong result or making a backward incompatible change?

Fixing the `wrong result`, not necessarily wrong but, it worked well for the previous definition cause it was limited.

>On the contrary, we should investigate if it is easy to build the definition of Feller/Comtet/Knuth on top of this. I suppose that it can be built with a simple piecewise wrapper, and this can be documented.

tbh I'm not exactly sure which implementation this is, cause FCK doesn't work for negative integers and JRL uses a limiting gamma value 🤷‍♂️.
our's is consistent with the version mathematica uses tho

**Comment by smichr** [comment]
> Should this be considered as fixing the wrong result or making a backward incompatible change?

As a suggestion for the future, it would be better to raise NotImplemented for an unhandled range of values (in this case k < 0) rather than making an arbitrary value "for convenience" (as was the previous behavior).

**Comment by asmeurer** [comment]
It's not a wrong result. The different definitions are all valid. There's no unique way of extending the binomial to negative arguments, so you have to define how to do it. See the discussions in https://github.com/sympy/sympy/issues/14577

**Comment by asmeurer** [comment]
> As a suggestion for the future, it would be better to raise NotImplemented for an unhandled range of values (in this case k < 0) rather than making an arbitrary value "for convenience" (as was the previous behavior).

The existing definition is not a random thing that was chosen "for convenience". It's a well established definition that has mathematical backing. See https://github.com/sympy/sympy/issues/14577 and the John Cook blog post. 

I don't think "the values are all 0" should be considered inherently a bad thing. That's what falls out of the limiting definition of gammas. One could just as well ask why factorials of negative integers are infinite, when it could be more "useful" for them to be some finite value. But the gamma function has a very clean mathematical definition, and nice properties (like being the only log-convex extension of factorial), and it just happens to have poles at negative integers. So we should be focused on which definition has the better inherent properties, not just which has values that seem nicer at a glance because they aren't 0.

**Comment by smichr** [comment]
> not a random thing that was chosen "for convenience"

I am just quoting the documentation (which may be wrong). Whereas `factorial` says

> By convention (consistent with the gamma function and the 
> binomial coefficients), factorial of a negative integer is complex infinity.
> 

`binomial` says

> For the sake of convenience for negative integer `k` this function will return zero no matter what 
> valued is the other argument.

And that automatically means the relationship `(n,k) = (n, n - k)` is being ignored since `(3,-2)` could be `(3, 5)`. But I really don't know much about these things and will rely on @iammosespaulr for the heavy lifting here (or anyone else that wants to help).

**Comment by asmeurer** [comment]
binomial(3, 5) is 0.

**Comment by iammosespaulr** [comment]
I think `binomial(-3 , -5)` is more appropriate 😅 .
`binomial(-3, 2)` is 6 and now `binomial(-3 , -5)` is also 6. earlier it used to be zero

> binomial(3, 5) is 0.

**Comment by iammosespaulr** [comment]
and @asmeurer I've got an idea about the gamma rewrite. One that is consistent with all integers negative and positive.
Yes gamma is infinity for any integers other than  <b><span>&#8469;</span> </b>.
I propose doing something like this.
Screen Shot 2020-04-23 at 2 04 06 AM
the version I just came up with seems to work flawlessly for both negative and positive ints.
@smichr  if this works out, we wouldn't need to use conditionals in `rewrite_gamma`

**Comment by asmeurer** [comment]
This is the sort of thing that should be documented (even if we keep the current definition it should be documented). The definition of binomial(n, k) for 1 <= k <= n integer is completely unambiguous, because its defined combinatorially (the number of ways of choosing k elements from n, or the size of the set {A | card(A)=k and x ⊆ {1, ..., n}}). The extension for k > n or k = 0 follows from the common conventions on counting. The ambiguity comes when n or k are negative integers, or more generally when they are arbitrary complex numbers. You can't define those cases combinatorially, so you have to extend the base definition somehow. You can use n!/(k!(n-k)!), and use gammas instead of factorials, but the issue is that for negative integers, the factorial is infinite, so you have to take limits. But there are two variables, so the limit isn't unique (it depends on what order you take limits). 

> I am just quoting the documentation (which may be wrong). Whereas factorial says

I see. That's a poorly chosen phrase in the documentation. There is a motivation for the definition. It just isn't unique in the same way factorial is. There aren't strong arguments to make factorial(-n) anything other than infinity. The only factorial identity is n! = n*(n-1)!, which works with the definition (0! = 1 = 0*oo is an indeterminate), but it does work out with the gamma extension, as well as the combinatorial definition of binomial. But for binomial there are arguments for both definitions, because some identities work for one and some for the other. And there isn't a strong combinatorial reason to choose one over the other, at least that I'm aware of (if there is let me know).

**Comment by smichr** [comment]
> I think `binomial(-3 , -5)` is more appropriate

yes...that's what I meant. Ugh. I should just stay out of this.

**Comment by iammosespaulr** [comment]
> And there isn't a strong combinatorial reason to choose one over the other, at least that I'm aware of (if there is let me know).

@asmeurer the reason why I advocate for -ve int n's and -k's are because.  

Take the example ( y + x ) <sup>-3</sup>
( I'm setting y = 1 for simplicity )

[(1 + x)<sup>-3</sup> = 1 - 3 x<sup>1</sup> + 6 x<sup>2</sup> - 10 x<sup>3</sup> + 15 x<sup>4</sup> + O(x<sup>5</sup>) (Taylor series)
(1 + x)<sup>-3</sup> = (1/x)<sup>3</sup> - 3/x<sup>4</sup> + 6/x<sup>5</sup> - 10/x<sup>6</sup> + O((1/x)<sup>7</sup>) (Laurent series)  ](http://www.wolframalpha.com/input/?i=(1%20%2B%20x)%5E%20-3)

`binomial(-3, 3)= 10`, `binomial(-3, 4) = 15` ,`binomial(-3, -5)= 6`, `binomial(-3, -6) = 10` and so on

so `binomial(n, k)` can be interpreted as the coefficient of the term with power `k` in a polynomial with two variables of order `n`

I think this beautifully captures the essence of having binomial support all integers. ( this interpretation works for complex numbers too )
This also easily extends to multinomials.

**Comment by asmeurer** [comment]
> so binomial(n, k) can be interpreted as the coefficient of the term with power k in a polynomial with two variables of order n

https://github.com/sympy/sympy/issues/14577#issuecomment-379046095

**Comment by iammosespaulr** [comment]
@asmeurer also, extending binomials to -ve integers gives `multinomial` a whole new level of symmetry

#### the order of the arguments shouldn't matter in multinomial !
---

#### With extension to negative ints...
```python3
binomial(n , k) = multinomial(k , n - k) = multinomial(n - k , k) = binomial(n , n - k)
```
 i.e `multinomial(i , j) == multinomial(j , i)`  

---

#### Without extension
```python3
binomial(n, k) == multinomial(k, n - k)
e.g
binomial(-3  , -5) == 0 == multinomial(-5 , 2)
binomial(-3 , 2) == 6 == multinomial(2, -5)
so
multinomial(i , j) != multinomial(j , i)
```

---

[This is just wrong! cause `multinomial(i , j) == multinomial(j , i)`](http://www.wolframalpha.com/input/?i=multinomial(i%20%2C%20j)%20%3D%3D%20multinomial(j%20%2C%20i))

**Comment by czgdp1807** [comment]
Any news on this? @smichr

**Review by asmeurer — CHANGES_REQUESTED** [review]
I'd like to see documentation on this before this is merged. https://github.com/sympy/sympy/pull/19158#issuecomment-617364174

**Comment by iammosespaulr** [comment]
> Another thing that I'd like to see is documentation that discusses the definition, what the advantages are, alternate definitions (it should at least be clear that there are alternate definitions), and so on. It can go in the docstring, unless it ends up being really long, in which case it may be better as a separate document in the Sphinx docs.

It's almost done, I added definitions for the main implementation.
we worked out most of the kinks w.r.t rewriting as gamma, so now simplification works for both neg int's etc.
Working on docs for ☝️ .

**Comment by smichr** [comment]
> Any news on this? @smichr

work is continuing at #18960 -- the addition of multinomial is trivial by comparison so let's doo all the review there.

**Pull request closed without merging**